subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/IvyVariantComponentMetadataRulesIntegrationTest.groovy

```diff
@@ -0,0 +1,235 @@
+                        withVariant('customConf') { 
+                            attributes { attribute(a1, 'select') }
+                        }
+                    }
+                }
+            }
+        """
+
+        then:
+        fails 'checkDep'
+        failure.assertHasCause """Cannot choose between the following variants of org.test:projectA:1.0:
+  - compile
+  - customConf
+All of them match the consumer attributes:
+  - Variant 'compile' capability org.test:projectA:1.0:
+      - Unmatched attribute:
+          - Found org.gradle.status 'integration' but wasn't required.
+      - Compatible attribute:
+          - Required a1 'select' and found compatible value 'select'.
+  - Variant 'customConf' capability org.test:projectA:1.0:
+      - Unmatched attribute:
+          - Found org.gradle.status 'integration' but wasn't required.
+      - Compatible attribute:
+          - Required a1 'select' and found compatible value 'select'."""
+    }
+
+    def "local explicit configuration selection wins over attribute matching"() {
+        given:
+        repository {
+            'org.test:projectA:1.0'()
+        }
+
+        when:
+        buildFile << """
+            def a1 = Attribute.of('a1', String)
+            configurations.conf {
+                attributes { attribute(a1, 'select') }
+            }
+            dependencies {
+                conf group: 'org.test', name: 'projectA', version: '1.0', configuration: 'runtime'
+                components {
+                    withModule('org.test:projectA') {
+                        withVariant('compile') { 
+                            attributes { attribute(a1, 'select') }
+                        }
+                    }
+                }
+            }
+        """
+        repositoryInteractions {
+            'org.test:projectA:1.0' { expectResolve() }
+        }
+
+        then:
+        succeeds 'checkDep'
+        resolve.expectGraph {
+            root(':', ':test:') {
+                // explicit selection of 'runtime', attribute matching is not used
+                module("org.test:projectA:1.0:runtime")
+            }
+        }
+    }
+
+    def "attribute matching wins over published explicit configuration selection"() {
+        given:
+        repository {
+            'org.test:projectA:1.0' {
+                dependsOn group: 'org.test', artifact: 'projectB', version: '1.0', conf: '*->another'
+            }
+            'org.test:projectB:1.0'() {
+                variant 'another', [:]
+            }
+        }
+
+        when:
+        buildFile << """
+            def a1 = Attribute.of('a1', String)
+            configurations.conf {
+                attributes { attribute(a1, 'select') }
+            }
+            dependencies {
+                conf group: 'org.test', name: 'projectA', version: '1.0'
+                components {
+                    all {
+                        withVariant('compile') { 
```

Comment:
I do not understand how this applied to `org.test:projectA:1.0` produces a variant with a dependency. Where is that dependency coming from?

Reply:
I think I got it, you are effectively transforming Ivy configurations in variants if and only if a rule that touches attributes is applied to it.
So the dependency comes from the Ivy `compile` configuration, right?

Reply:
Exactly. That would be the feature we offer with this PR. You can transform pure _ivy configurations_ into _Gradle variants_ by adding attributes. Everything else is "copied" from the configurations.